### PR TITLE
Add cgroups metric collector for containers and pods

### PR DESCRIPTION
**Pull request by karan**
This collector will export metrics for `kubepods` cgroups. In the future, this can be extended to also support non-k8s controls (`system` label).

Tested on a COS VM (spot checked):

```
karangoel@gke-cos-rapid-cos-ctd-rapid-c38076b8-dpmx ~ $ curl -s localhost:20257/metrics  | grep -i ^cgroup | head
cgroup_cpu_periods_count{container_id="1f03b3c855d7875f4ec28bb0b4ff2aa1235f328acff79900d041f5a45e0d0d8d",pod_id="a44b8e4f-2a86-46df-8322-fe815e297d53",service_name=""} 0
cgroup_cpu_periods_count{container_id="30e20e68337dc15c40adda6108667712194134ca2a4971b15e928c249013a7fb",pod_id="0f2ee550-7fb8-4596-91cf-2c191f2e68c6",service_name=""} 0
cgroup_cpu_periods_count{container_id="32af853d2ae4a858fa5a1baa4ea4d9d654c52293cab5e5ca700d1250ea285818",pod_id="db2811955cb8619ad8871fc79c3e2e15",service_name=""} 0
cgroup_cpu_periods_count{container_id="5065d07579b2eee8c114d63752188384bdaac5d7251f69e06f19ead6dbd954d0",pod_id="a0247500-5e9f-4666-885e-3c2252dac380",service_name=""} 0
cgroup_cpu_periods_count{container_id="6bb0d4f69daf7666239a719c732e28fc171a27859e8b44b56580107a22b840bf",pod_id="a0247500-5e9f-4666-885e-3c2252dac380",service_name=""} 0
cgroup_cpu_periods_count{container_id="70dc2440a420a68db86cffcb9582046b85e693649694d2df37873b734441552b",pod_id="db2811955cb8619ad8871fc79c3e2e15",service_name=""} 0
cgroup_cpu_periods_count{container_id="7fce613225ef98af693f15b85b2fee02327bd27bf2de2d28c01d6890e0703bf1",pod_id="86b2d595-a118-4b82-8727-c310a4607f0f",service_name=""} 0
cgroup_cpu_periods_count{container_id="9bec7003b5674bc4c9a4185414b36bf883958190fab49ea27402fc81bf7dc478",pod_id="86b2d595-a118-4b82-8727-c310a4607f0f",service_name=""} 0
cgroup_cpu_periods_count{container_id="a8e6823af7267312af28108ce688af88a34cf02023d7dffac3686d7b2a3e05b4",pod_id="86b2d595-a118-4b82-8727-c310a4607f0f",service_name=""} 0
cgroup_cpu_periods_count{container_id="ae7abd5f0b6bf15bd784585d92444cab588953c297a72357f4b92c80641e7672",pod_id="86b2d595-a118-4b82-8727-c310a4607f0f",service_name=""} 0
karangoel@gke-cos-rapid-cos-ctd-rapid-c38076b8-dpmx ~ $ curl -s localhost:20257/metrics  | grep -i ^cgroup | tail
cgroup_pids_current_count{container_id="6bb0d4f69daf7666239a719c732e28fc171a27859e8b44b56580107a22b840bf",pod_id="a0247500-5e9f-4666-885e-3c2252dac380",service_name=""} 9
cgroup_pids_current_count{container_id="70dc2440a420a68db86cffcb9582046b85e693649694d2df37873b734441552b",pod_id="db2811955cb8619ad8871fc79c3e2e15",service_name=""} 1
cgroup_pids_current_count{container_id="7fce613225ef98af693f15b85b2fee02327bd27bf2de2d28c01d6890e0703bf1",pod_id="86b2d595-a118-4b82-8727-c310a4607f0f",service_name=""} 8
cgroup_pids_current_count{container_id="9bec7003b5674bc4c9a4185414b36bf883958190fab49ea27402fc81bf7dc478",pod_id="86b2d595-a118-4b82-8727-c310a4607f0f",service_name=""} 11
cgroup_pids_current_count{container_id="a8e6823af7267312af28108ce688af88a34cf02023d7dffac3686d7b2a3e05b4",pod_id="86b2d595-a118-4b82-8727-c310a4607f0f",service_name=""} 9
cgroup_pids_current_count{container_id="ae7abd5f0b6bf15bd784585d92444cab588953c297a72357f4b92c80641e7672",pod_id="86b2d595-a118-4b82-8727-c310a4607f0f",service_name=""} 1
cgroup_pids_current_count{container_id="b40939f447a3be5b043186e31fa15605a6aa0d4df290f934f9c38467645d4510",pod_id="a44b8e4f-2a86-46df-8322-fe815e297d53",service_name=""} 1
cgroup_pids_current_count{container_id="cfa27cc6038df0571bacdc61ac1e5baae52e686b9346b025268657d532c8ee06",pod_id="86b2d595-a118-4b82-8727-c310a4607f0f",service_name=""} 9
cgroup_pids_current_count{container_id="edd74f8bd5cfa99371d6ec4bf0f6f098082d0e414d04a08895317be9159a3162",pod_id="a44b8e4f-2a86-46df-8322-fe815e297d53",service_name=""} 5
cgroup_pids_current_count{container_id="f59f8d0a54951c307b46678073504d52e1b62117389cb72680884f9b6d4c0945",pod_id="0f2ee550-7fb8-4596-91cf-2c191f2e68c6",service_name=""} 11
```

### CODE REVIEW

File: pkg/systemstatsmonitor/labels.go [hunk 1]
```diff
@@ -45,3 +45,21 @@ const cpuLabel = "cpu"
...
+const containerIDLabel = "container_id"
+
+// serviceNameLabel labels the systemd service
+const serviceNameLabel = "service_name"
+
+// userLabel labels if the feature is for user space
+const userLabel = "user"
+
+// systemLabel labels if the feature is for system space
+const systemLabel = "user"
```
Comment (vteratipally) [inline pkg/systemstatsmonitor/labels.go:62]: nit: systemLabel = "system"?

File: pkg/systemstatsmonitor/cgroup_collector.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2020 The Kubernetes Authors All rights reserved.
```
Comment (vteratipally) [inline pkg/systemstatsmonitor/cgroup_collector.go:2]: 2021

Comment (vteratipally) [inline pkg/systemstatsmonitor/cgroup_collector.go:2]: 2021

File: pkg/util/cgroup/cgroup.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
+package cgroup
```
Comment (vteratipally) [inline pkg/util/cgroup/cgroup.go:1]: License header

File: pkg/systemstatsmonitor/labels.go [hunk 4]
```diff
@@ -45,3 +45,21 @@ const cpuLabel = "cpu"
...
+// serviceNameLabel labels the systemd service
+const serviceNameLabel = "service_name"
+
+// userLabel labels if the feature is for user space
+const userLabel = "user"
+
+// systemLabel labels if the feature is for system space
+const systemLabel = "system"
+
+// typeLabel labels the memory type
```
Comment (vteratipally) [inline pkg/systemstatsmonitor/labels.go:64]: As type can be anything here, can we change it to just type?

**Review by vteratipally — APPROVED** [review]

**Comment by Random-Liu** [comment]
Why would we need per container cgroup stats from NPD?

**Comment by karan** [comment]
The main reason is that we cannot rely on cAdvisor or the kubelet alone in case the kubelet itself is unhealthy. We've seen that happen many, many times and we lose out on container metrics altogether.

As it is implemented here, NPD doesn't actually rely on "containers" - it's container runtime agnostic. It's merely reading the cgroup fs - today for `kubepods` and in the future for system units as well.

**Comment by karan** [comment]
The main reason is that we cannot rely on cAdvisor or the kubelet alone in case the kubelet itself is unhealthy. We've seen that happen many, many times and we lose out on container metrics altogether.

As it is implemented here, NPD doesn't actually rely on "containers" - it's container runtime agnostic. It's merely reading the cgroup fs - today for `kubepods` and in the future for system units as well.

**Comment by Random-Liu** [comment]
Based on offline discussion, we will not add pod/container level metrics support into NPD for now.

**Comment by Random-Liu** [comment]
As is discussed, we won't add pod/container level cgroup metrics in NPD.

**Pull request closed without merging**